zeus/monitor/carbon.py

```diff
@@ -0,0 +1,448 @@
+        if len(self.current_keys) == 1:
+            self.polling_process = self.context.Process(
+                target=_polling_process,
+                args=(
+                    self.command_q,
+                    self.finished_q,
+                    self.zeus_monitor.gpu_indices,
+                    self.zeus_monitor.cpu_indices,
+                    self.carbon_intensity_provider,
+                ),
+            )
+            self.polling_process.start()
+
+        # start subwindows
+        self.command_q.put((self.Op.BEGIN, key))
+
+    def end_window(
+        self, key: str, sync_execution: bool = True
+    ) -> CarbonEmissionMeasurement:
+        """End a measurement window and return the time, energy consumption, and carbon emission.
+
+        Args:
+                        key: Name of an active measurement window.
+                        sync_execution: Whether to wait for asynchronously dispatched computations
+                                        to finish before starting the measurement window. For instance, PyTorch
+                                        and JAX will run GPU computations asynchronously, and waiting them to
+                                        finish is necessary to ensure that the measurement window captures all
+                                        and only the computations dispatched within the window.
+        """
+        # check if begin_window has been called with key before
+        if key not in self.current_keys:
+            raise ValueError(f"Measurement window '{key}' does not exist")
+
+        # Synchronize execution (e.g., cudaSynchronize) to freeze at the right time.
+        if sync_execution and self.zeus_monitor.gpu_indices:
+            sync_execution_fn(self.zeus_monitor.gpu_indices, sync_with=self.sync_with)
+
+        # end window
+        self.command_q.put((self.Op.END, key))
+
+        # continue fetching until you find the key you have received
+        while key not in self.finished_keys:
+            try:
+                (
+                    retrieved_key,
+                    gpu_carbon_emission,
+                    cpu_carbon_emission,
+                ) = self.finished_q.get_nowait()
+                self.finished_keys[retrieved_key] = (
+                    gpu_carbon_emission,
+                    cpu_carbon_emission,
+                )
+                self.current_keys.remove(key)
+            except queue.Empty:
+                pass
+        overall_measurement = self.zeus_monitor.end_window(key)
+
+        measurement = CarbonEmissionMeasurement(
+            time=overall_measurement.time,
+            gpu_energy=overall_measurement.gpu_energy,
+            cpu_energy=overall_measurement.cpu_energy,
+            dram_energy=overall_measurement.dram_energy,
+            gpu_carbon_emission=self.finished_keys[key][0],
+            cpu_carbon_emission=self.finished_keys[key][1] or None,
```

Comment:
What about DRAM?

Reply:
Also, in the test file, the mock `dram_energy` was empty. That won't cover any of the code paths related to DRAM.

Reply:
Just to clarify: add a dram_carbon_emission too and test that?

Reply:
Yes, please.